### PR TITLE
Fix to use armeria client's uri for Retrofit's base URL

### LINKED ISSUE #376: The raw okhttp3.Request object always has 0.0.0.0 authority when armeria-retrofit2 is used
When you use armeria-retrofit2 and receive a response from Retrofit2:

```java
        // From ArmeriaCallFactoryTest.java
        Response<Void> response = service.postForm("Cony", 26).get();
        assertThat(response.isSuccessful());
        System.err.println(response.raw().request().url());
```

.. it will print the following URL that contains a misleading 0.0.0.0 rather than the actual hostname:

```
http://0.0.0.0/postForm
```

This can make a user think the request was sent to a wrong host.

### PR DESCRIPTION
**Pull request by imasahiro**
Motivations:
- Currently, a response of armeria-retrofit returns wrong uri like `"http://127.0.0.1/"` and it can make a user confusion

Modifications:
- Fix to use armeria client's uri instead of dummy uri for Retrofit's base URL
- Use AbstractServerTest at ArmeriaCallFactoryTest
- Fix to use try-with resource for Buffer due to Buffer is AutoCloseable

Results:
- Fixes #376

### CODE REVIEW

File: retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaRetrofit.java [hunk 1]
```diff
@@ -37,17 +43,25 @@
...
-                .baseUrl(UNUSED_URL)
+                .baseUrl(convertToSerializationFormatRemovedUri(httpClient.uri()))
                 .callFactory(new ArmeriaCallFactory(httpClient));
     }
 
+    @VisibleForTesting
+    static String convertToSerializationFormatRemovedUri(URI uri) {
+        String uriStr = uri.toString();
+        int schemeOffset = uriStr.indexOf("://");
+        Preconditions.checkArgument(schemeOffset >= 0, "uri does not contains the scheme component.");
```
Comment (trustin) [inline retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaRetrofit.java]: contains -> contain

File: retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaRetrofit.java [hunk 2]
```diff
@@ -37,17 +43,25 @@
...
     public static Retrofit.Builder builder(HttpClient httpClient) {
         return new Retrofit.Builder()
-                .baseUrl(UNUSED_URL)
+                .baseUrl(convertToSerializationFormatRemovedUri(httpClient.uri()))
                 .callFactory(new ArmeriaCallFactory(httpClient));
     }
 
+    @VisibleForTesting
+    static String convertToSerializationFormatRemovedUri(URI uri) {
+        String uriStr = uri.toString();
```
Comment (trustin) [inline retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaRetrofit.java]: You could just call `uri.getScheme()` or `uri.getRawScheme()` and check the returned value is `null`?

File: retrofit2/src/test/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaCallFactoryTest.java [hunk 3]
```diff
@@ -309,4 +287,15 @@ public void onFailure(Call<Pojo> call, Throwable t) {
         pojoCall.cancel();
         assertThat(failCount.intValue()).isZero();
     }
+
+    @Test
+    public void respectsHttpClientUri() throws Exception {
```
Comment (kojilin) [inline retrofit2/src/test/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaCallFactoryTest.java]: Add test for EndpointGroupRegistry?

Comment (trustin) [inline retrofit2/src/test/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaCallFactoryTest.java]: Sounds like a good suggestion. I wonder what will happen when the authority is 'group:someGroup'.

Comment (imasahiro) [inline retrofit2/src/test/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaCallFactoryTest.java]: Right, need to fix. Actually, when I try to use EndpointGroup, it does not work.
```
java.lang.IllegalArgumentException: Illegal URL: http://group:group/
	at retrofit2.Retrofit$Builder.baseUrl(Retrofit.java:437)
	at com.linecorp.armeria.client.http.retrofit2.ArmeriaRetrofit.builder(ArmeriaRetrofit.java:53)
```
I need to add some changes to ArmeriaCallSubscriber to update okhttp's url

**Comment by trustin** [comment]
Thanks, @imasahiro !

**Pull request merged**